Compile error due to the wrong literal data type

<!--
Thank you for filing a bug report! 🐛 Please provide a short summary of the bug,
along with any information you feel relevant to replicating the bug.
-->

I tried this code:

```rust
fn main() {
    let array = [0x42u8; 10];
    for b in &array {
        let lo = b & 0xf;
        let hi = (b >> 4) & 0xf;
    }
}
```
[Playground](https://play.rust-lang.org/?version=stable&mode=debug&edition=2018&gist=7558b47acf4942b543c5d8bf5a2808cb)

I expected to see this happen: it should be compiled

Instead, this happened: 
```
error[E0277]: no implementation for `u8 & i32`
 --> src/main.rs:5:27
  |
5 |         let hi = (b >> 4) & 0xf;
  |                           ^ no implementation for `u8 & i32`
  |
  = help: the trait `std::ops::BitAnd<i32>` is not implemented for `u8`
```

### Meta
<!--
If you're using the stable version of the compiler, you should also check if the
bug also exists in the beta or nightly versions.
-->

`rustc --version --verbose`:
```
rustc 1.41.0 (5e1a79984 2020-01-27)
binary: rustc
commit-hash: 5e1a799842ba6ed4a57e91f7ab9435947482f7d8
commit-date: 2020-01-27
host: x86_64-unknown-linux-gnu
release: 1.41.0
LLVM version: 9.0
```
The same happens on beta and nightly.

<!--
Include a backtrace in the code block by setting `RUST_BACKTRACE=1` in your
environment. E.g. `RUST_BACKTRACE=1 cargo build`.
-->
<details><summary>Backtrace</summary>
<p>

```
Compiling playground v0.0.1 (/playground)
error[E0277]: no implementation for `u8 & i32`
 --> src/main.rs:5:27
  |
5 |         let hi = (b >> 4) & 0xf;
  |                           ^ no implementation for `u8 & i32`
  |
  = help: the trait `std::ops::BitAnd<i32>` is not implemented for `u8`

error: aborting due to previous error

For more information about this error, try `rustc --explain E0277`.
error: could not compile `playground`.

To learn more, run the command again with --verbose.
```

</p>
</details>

--- a/src/test/ui/const-generics/issues/issue-75299.rs
+++ b/src/test/ui/const-generics/issues/issue-75299.rs
@@ -1,0 +1,11 @@
+// compile-flags: -Zmir-opt-level=3
+// run-pass
+
+#![feature(const_generics)]
+#![allow(incomplete_features)]
+fn main() {
+    fn foo<const N: usize>() -> [u8; N] {
+        [0; N]
+    }
+    let _x = foo::<1>();
+}

--- a/src/test/ui/issues/issue-68951.rs
+++ b/src/test/ui/issues/issue-68951.rs
@@ -1,0 +1,9 @@
+// check-pass
+
+fn main() {
+    let array = [0x42u8; 10];
+    for b in &array {
+        let lo = b & 0xf;
+        let hi = (b >> 4) & 0xf;
+    }
+}

--- a/src/test/ui/pattern/issue-66501.rs
+++ b/src/test/ui/pattern/issue-66501.rs
@@ -1,0 +1,12 @@
+// check-pass
+
+#![allow(unreachable_patterns)]
+
+fn main() {
+    const CONST: &[Option<()>; 1] = &[Some(())];
+    match &[Some(())] {
+        &[None] => {}
+        CONST => {}
+        &[Some(())] => {}
+    }
+}

--- a/src/test/ui/pattern/issue-72565.rs
+++ b/src/test/ui/pattern/issue-72565.rs
@@ -1,0 +1,8 @@
+const F: &'static dyn PartialEq<u32> = &7u32;
+
+fn main() {
+    let a: &dyn PartialEq<u32> = &7u32;
+    match a {
+        F => panic!(), //~ ERROR: `&dyn PartialEq<u32>` cannot be used in patterns
+    }
+}

--- a/src/test/ui/pattern/issue-72565.stderr
+++ b/src/test/ui/pattern/issue-72565.stderr
@@ -1,0 +1,8 @@
+error: `&dyn PartialEq<u32>` cannot be used in patterns
+  --> $DIR/issue-72565.rs:6:9
+   |
+LL |         F => panic!(),
+   |         ^
+
+error: aborting due to previous error
+

--- a/src/test/ui/type-alias-impl-trait/issue-74244.rs
+++ b/src/test/ui/type-alias-impl-trait/issue-74244.rs
@@ -1,0 +1,20 @@
+#![feature(type_alias_impl_trait)]
+
+trait Allocator {
+    type Buffer;
+}
+
+struct DefaultAllocator;
+
+impl<T> Allocator for DefaultAllocator {
+    //~^ ERROR: the type parameter `T` is not constrained
+    type Buffer = ();
+}
+
+type A = impl Fn(<DefaultAllocator as Allocator>::Buffer);
+
+fn foo() -> A {
+    |_| ()
+}
+
+fn main() {}

--- a/src/test/ui/type-alias-impl-trait/issue-74244.stderr
+++ b/src/test/ui/type-alias-impl-trait/issue-74244.stderr
@@ -1,0 +1,9 @@
+error[E0207]: the type parameter `T` is not constrained by the impl trait, self type, or predicates
+  --> $DIR/issue-74244.rs:9:6
+   |
+LL | impl<T> Allocator for DefaultAllocator {
+   |      ^ unconstrained type parameter
+
+error: aborting due to previous error
+
+For more information about this error, try `rustc --explain E0207`.

